## [作業重點]
確保你了解隨機森林模型中每個超參數的意義，並觀察調整超參數對結果的影響

## 作業

1. 試著調整 RandomForestClassifier(...) 中的參數，並觀察是否會改變結果？
2. 改用其他資料集 (boston, wine)，並與回歸模型與決策樹的結果進行比較

In [1]:
from sklearn import datasets, metrics
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

C:\Users\USER\Anaconda3\lib\site-packages\sklearn\ensemble\weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.
  from numpy.core.umath_tests import inner1d


## 使用 20 顆樹，每棵樹的最大深度為 4：

In [3]:
# 讀取boston資料集
wine = datasets.load_wine()

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(wine.data, wine.target, test_size=0.25, random_state=4)

# 建立模型 (使用 20 顆樹，每棵樹的最大深度為 4)
clf = RandomForestClassifier(n_estimators=20, max_depth=4)

# 訓練模型
clf.fit(x_train, y_train)

# 預測測試集
y_pred = clf.predict(x_test)

In [4]:
acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)

Accuracy:  0.9777777777777777


In [6]:
print(wine.feature_names)

['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']


In [7]:
print("Feature importance: ", clf.feature_importances_)

Feature importance:  [0.14032002 0.01987009 0.00774303 0.02934035 0.04437088 0.00927024
 0.21267335 0.01783473 0.01864012 0.19445515 0.01879632 0.14126561
 0.14542011]


## 使用 40 顆樹，每棵樹的最大深度為 4：

In [8]:
# 讀取boston資料集
wine = datasets.load_wine()

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(wine.data, wine.target, test_size=0.25, random_state=4)

# 建立模型
clf = RandomForestClassifier(n_estimators=40, max_depth=4)

# 訓練模型
clf.fit(x_train, y_train)

# 預測測試集
y_pred = clf.predict(x_test)

In [9]:
acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)

Accuracy:  1.0


In [10]:
print(wine.feature_names)

['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']


In [11]:
print("Feature importance: ", clf.feature_importances_)

Feature importance:  [0.15132864 0.01782648 0.01251362 0.04032581 0.02621662 0.05497414
 0.12506472 0.00876627 0.02924671 0.19355321 0.09326201 0.11999904
 0.12692275]


## 使用 40 顆樹，每棵樹的最大深度為 16：

In [12]:
# 讀取boston資料集
wine = datasets.load_wine()

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(wine.data, wine.target, test_size=0.25, random_state=4)

# 建立模型 (使用 20 顆樹，每棵樹的最大深度為 4)
clf = RandomForestClassifier(n_estimators=40, max_depth=16)

# 訓練模型
clf.fit(x_train, y_train)

# 預測測試集
y_pred = clf.predict(x_test)

In [13]:
acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)

Accuracy:  0.9777777777777777


In [14]:
print(wine.feature_names)

['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']


In [15]:
print("Feature importance: ", clf.feature_importances_)

Feature importance:  [0.09640456 0.04125677 0.01384849 0.05548233 0.0298252  0.04665905
 0.14653321 0.01095386 0.02373455 0.20283082 0.06389803 0.15308136
 0.11549176]


## 使用 20 顆樹，每棵樹的最大深度為 8：

In [16]:
# 讀取boston資料集
wine = datasets.load_wine()

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(wine.data, wine.target, test_size=0.25, random_state=4)

# 建立模型 (使用 20 顆樹，每棵樹的最大深度為 4)
clf = RandomForestClassifier(n_estimators=20, max_depth=8)

# 訓練模型
clf.fit(x_train, y_train)

# 預測測試集
y_pred = clf.predict(x_test)

In [17]:
acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)

Accuracy:  0.9777777777777777


In [18]:
print(wine.feature_names)

['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']


In [19]:
print("Feature importance: ", clf.feature_importances_)

Feature importance:  [0.16814852 0.06105098 0.00804527 0.04331974 0.05307555 0.03838308
 0.10214307 0.01225667 0.02238568 0.18403966 0.0404082  0.16244779
 0.10429579]


## 使用 20 顆樹，每棵樹的最大深度為 2：

In [20]:
# 讀取boston資料集
wine = datasets.load_wine()

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(wine.data, wine.target, test_size=0.25, random_state=4)

# 建立模型 (使用 20 顆樹，每棵樹的最大深度為 4)
clf = RandomForestClassifier(n_estimators=20, max_depth=2)

# 訓練模型
clf.fit(x_train, y_train)

# 預測測試集
y_pred = clf.predict(x_test)

In [21]:
acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)

Accuracy:  0.9555555555555556


In [22]:
print(wine.feature_names)

['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']


In [23]:
print("Feature importance: ", clf.feature_importances_)

Feature importance:  [0.1767701  0.02481086 0.         0.02850617 0.01835629 0.04479562
 0.17474886 0.         0.01981717 0.17196318 0.09439285 0.11778041
 0.12805849]


## 使用 40 顆樹，每棵樹的最大深度為 8：

In [24]:
# 讀取boston資料集
wine = datasets.load_wine()

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(wine.data, wine.target, test_size=0.25, random_state=4)

# 建立模型 (使用 20 顆樹，每棵樹的最大深度為 4)
clf = RandomForestClassifier(n_estimators=20, max_depth=2)

# 訓練模型
clf.fit(x_train, y_train)

# 預測測試集
y_pred = clf.predict(x_test)

In [25]:
acc = metrics.accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)

Accuracy:  1.0


In [26]:
print(wine.feature_names)

['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']


In [27]:
print("Feature importance: ", clf.feature_importances_)

Feature importance:  [0.23928549 0.         0.02089329 0.0050182  0.         0.01790796
 0.19330313 0.00202968 0.03137963 0.10786781 0.08684583 0.19814353
 0.09732546]


# 結論：
使用 20 顆樹，每棵樹的最大深度為 4：Accuracy: 0.9777777777777777<br>
使用 20 顆樹，每棵樹的最大深度為 8：Accuracy: 0.9777777777777777<br>
使用 20 顆樹，每棵樹的最大深度為 2：Accuracy: 0.9555555555555556<br>
使用 40 顆樹，每棵樹的最大深度為 4：Accuracy: 1.0<br>
使用 40 顆樹，每棵樹的最大深度為 8：Accuracy: 1.0<br>
使用 40 顆樹，每棵樹的最大深度為 16：Accuracy: 0.9777777777777777<br>

同樣數量20顆的樹，深度越少準確率越低，到達一定深度後準確率就固定<br>
但是同樣數量40顆的樹，樹的深度越大，準確率有下降的趨勢<br>


# 參考解答：

In [1]:
from sklearn.datasets import load_boston, load_wine
from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score, accuracy_score

C:\Users\USER\Anaconda3\lib\site-packages\sklearn\ensemble\weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.
  from numpy.core.umath_tests import inner1d


In [2]:
# 讀取 Boston 資料
boston = load_boston()

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(boston.data, boston.target, test_size=0.1, random_state=4)

# 建立一個線性回歸模型
regr = RandomForestRegressor()

# 將訓練資料丟進去模型訓練
regr.fit(x_train, y_train)

# 將測試資料丟進模型得到預測結果
y_pred = regr.predict(x_test)

In [3]:
print("Mean squared error: %.2f"
      % mean_squared_error(y_test, y_pred))

Mean squared error: 11.31


In [4]:
# 讀取 wine 資料
wine = load_wine()

# 切分訓練集/測試集
x_train, x_test, y_train, y_test = train_test_split(wine.data, wine.target, test_size=0.1, random_state=4)

# 建立一個羅吉斯回歸模型
regr = RandomForestClassifier()

# 將訓練資料丟進去模型訓練
regr.fit(x_train, y_train)

# 將測試資料丟進模型得到預測結果
y_pred = regr.predict(x_test)

In [5]:
acc = accuracy_score(y_test, y_pred)
print("Accuracy: ", acc)

Accuracy:  1.0
